### PR TITLE
created api post to find a best rate

### PR DESCRIPTION
**Pull request by rassl95**

### CODE REVIEW

File: src/main/java/com/github/hey_world_team/currency_converter/controller/rest/CentralBankApiController.java [hunk 1]
```diff
@@ -70,6 +76,19 @@ public ResponseEntity<Collection<History>> getAllHistory() {
...
+            @RequestParam("amount") BigDecimal amount,
+            @RequestParam("sourceCurrency") String sourceCurrency,
+            @RequestParam("costUSDT") BigDecimal costUSDT,
+            @RequestParam("costETH") BigDecimal costETH,
+            @RequestParam("costBTC") BigDecimal costBTC,
+            @RequestParam("targetCurrency") String targetCurrency) {
+
+        ConversionResponse result = conversionService.convertCurrency(amount, sourceCurrency,costUSDT, costETH, costBTC, targetCurrency);
+
+         return new ResponseEntity<>(result, HttpStatus.OK);
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/controller/rest/CentralBankApiController.java:90]: http 200 ok hardcoded

File: src/main/java/com/github/hey_world_team/currency_converter/model/ConversionResponse.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+import java.math.BigDecimal;
+import java.util.List;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@EqualsAndHashCode
+@ToString
+public class Response {
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/model/ConversionResponse.java:13]: have to  fix the name to BestConvertion

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+import java.util.*;
+
+@Service
+public class ConversionService {
+
+    private final Map<String, BigDecimal> cryptoRatesToUsd;
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:14]: add a comment something like a temporary decision for ...

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    private final Map<String, BigDecimal> cryptoRatesToRub;
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:15]: the same

Comment (dmitr2ish) [reply src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:15]: // TODO

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        cryptoRatesToUsd.put("ETH", new BigDecimal("1880")); // Курс ETH к USD
+        cryptoRatesToUsd.put("BTC", new BigDecimal("29800")); // Курс BTC к USD
+        //USD 10000
+        cryptoRatesToRub = new LinkedHashMap<>();
+        cryptoRatesToRub.put("USDT", new BigDecimal("92")); // Курс USDT к RUB
+        cryptoRatesToRub.put("ETH", new BigDecimal("173000")); // Курс ETH к RUB
+        cryptoRatesToRub.put("BTC", new BigDecimal("2456000")); // Курс BTC к RUB
+    }
+
+    public ConversionResponse convertCurrency(
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:31]: code style

File: src/main/java/com/github/hey_world_team/currency_converter/controller/rest/CentralBankApiController.java (hunk 1 above)
Comment (dmitr2ish) [reply src/main/java/com/github/hey_world_team/currency_converter/controller/rest/CentralBankApiController.java:90]: check the @ExceptionHandler and @AdviceController spring annotations

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+            // Если convertedAmount больше bestRate, обновляем bestRate
+            if (compareResult > 0) {
+                bestRate = convertedAmount;
+                bestCrypto = crypto;
+                response.setConversionPath(String.format("The best way from %s to %s across %s will be costs %s", sourceCurrency, targetCurrency, bestCrypto, bestRate));
+                response.setConvertedAmount(bestRate);
+            }
+            i++;
+        }
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:68]: wrap to private method findBestCryptoPrice

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            BigDecimal costETH,
+            BigDecimal costBTC,
+            String targetCurrency) {
+        ConversionResponse response = new ConversionResponse();
+
+        // Проверяем, соответствует ли sourceCurrency одному из значений USD, RUB или EURO
+        if (!Arrays.asList("USD", "RUB", "EUR").contains(sourceCurrency.toUpperCase())) {
+            throw new IllegalArgumentException("Invalid sourceCurrency. Only USD, RUB, or EUR are allowed.");
+        }
+        List<BigDecimal> cost = new ArrayList<>();
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:44]: costList

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+        cost.add(costUSDT);
+        cost.add(costETH);
+        cost.add(costBTC);
+
+        String bestCrypto;
+        BigDecimal bestRate = BigDecimal.ZERO;
+
+        int i = 0;
+        for (String crypto : cryptoRatesToUsd.keySet()) {
+            BigDecimal convertedAmount = amount.divide(cost.get(i), 4, RoundingMode.HALF_EVEN)
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:54]: 4 - bad practice - magic digit - should to wrap to variable

**Review by dmitr2ish — CHANGES_REQUESTED** [review]
nice job, should be little bit fix

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+    //temporary solution for storing data about cryptocurrencies
+    private final Map<String, BigDecimal> cryptoRatesToUsd;
+    private final Map<String, BigDecimal> cryptoRatesToRub;
+    private final Map<String, BigDecimal> cryptoRatesToEur;
+
+    @Autowired
+    public ConversionService(CurrencyService service) {
+        // Rub 100000
+        cryptoRatesToUsd = new LinkedHashMap<>();
+        cryptoRatesToUsd.put("USDT", new BigDecimal("1.01")); // Курс USDT к USD
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:24]: need to translate russian comments

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+                                          BigDecimal costETH,
+                                          BigDecimal costBTC,
+                                          String targetCurrency) {
+        BestConversion response;
+        List<BigDecimal> costList = new ArrayList<>();
+        costList.add(costUSDT);
+        costList.add(costETH);
+        costList.add(costBTC);
+
+        // Проверяем, соответствует ли sourceCurrency одному из значений USD, RUB или EURO
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:51]: translate russian comments

File: src/main/java/com/github/hey_world_team/currency_converter/model/BestConversion.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+
+import java.math.BigDecimal;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@EqualsAndHashCode
+@ToString
+public class BestConversion {
+    private BigDecimal convertedAmount;
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/model/BestConversion.java:13]: what is mean this value?

File: src/main/java/com/github/hey_world_team/currency_converter/model/BestConversion.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+import lombok.*;
+
+import java.math.BigDecimal;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@EqualsAndHashCode
+@ToString
+public class BestConversion {
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/model/BestConversion.java:12]: java doc comment absent

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    public BestConversion convertCurrency(BigDecimal amount,
+                                          String sourceCurrency,
+                                          BigDecimal costUSDT,
+                                          BigDecimal costETH,
+                                          BigDecimal costBTC,
+                                          String targetCurrency) {
+        BestConversion response;
+        List<BigDecimal> costList = new ArrayList<>();
```
Comment (dmitr2ish) [inline src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:46]: use hashmap here, key - currency, value - cost

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java (hunk 9 above)
Comment (dmitr2ish) [reply src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:24]: also, i think  we don't need this part

if the cost of cryptocurrencies in source currency

our path next:

we have source currency -> crypto cost of source currency -> target currency in the request

so, we need to take from the repository cost of 1 target currency by  source currency

also, we need to prepare the correct cost for every crypto by the amount of source currency

then we need to compare their

**Review by dmitr2ish — CHANGES_REQUESTED** [review]
doesn't work

File: src/main/java/com/github/hey_world_team/currency_converter/model/BestConversion.java (hunk 11 above)
Comment (rassl95) [reply src/main/java/com/github/hey_world_team/currency_converter/model/BestConversion.java:13]: convertedAmount- the best rate you can get

File: src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java (hunk 13 above)
Comment (rassl95) [reply src/main/java/com/github/hey_world_team/currency_converter/service/ConversionService.java:46]: I don't understand how to use map  in double cycle

**Review by dmitr2ish — APPROVED** [review]

**Pull request merged**